src/tpm2.rs

```diff
@@ -0,0 +1,955 @@
+                            ),
+                            false => format!(
+                                "tpm2_evictcontrol -a o -c {} -P {}",
+                                hex::encode(curr_handle),
+                                owner_pw,
+                            ),
+                        };
+
+                        self.run(cmd, None)?;
+                        self.update_tpmdata("ek_handle", "")?;
+                        self.update_tpmdata("ek_pw", "")?;
+                    }
+                    Ok(())
+                })?;
+        }
+
+        if owner_pw.is_empty() {
+            owner_pw = self.random_password(20)?;
+            self.update_tpmdata("owner_pw", &owner_pw)?;
+        }
+
+        let ek_pw = self.random_password(20)?;
+        let tf = NamedTempFile::new()?;
+        let tf_path = self.temp_file_get_path(&tf)?;
+
+        let cmd = match self.legacy_tools {
+            true => format!(
+                "tpm2_getpubek -H 0x81010007 -g {} -f {} -P {} -o {} -e {}",
+                asym_alg, tf_path, ek_pw, owner_pw, owner_pw
+            ),
+            false => format!(
+                "tpm2_createek -c - -G {} -p {} -P {} -o {} -e {}",
+                asym_alg, tf_path, ek_pw, owner_pw, owner_pw
+            ),
+        };
+
+        self.run(
+            cmd,
+            Some(vec![tf_path]),
+        )
+        .and_then(|(ret_out, ret_f)| {
+            match self.legacy_tools {
+                true => self.update_tpmdata("ek_handle", "0x81010007")?,
+                false => {
+                    let ret_out_map: Value = serde_yaml::from_str(&ret_out)?;
+                    match ret_out_map["persistent-handle"].as_str(){
+                        None => self.update_tpmdata("ek_handle", "")?,
+                        Some(handle) => {
+                            self.update_tpmdata("ek_handle", &handle)?
+                        }
+                    }
+                },
+            }
+
+            ret_f.get(tf_path).map_or_else(
+                || {
+                    Err(KeylimeTpmError::new_tpm_rust_error(
+                        "tpm2_readpublic fail, ek key is missing is output file.",
+                    ))
+                },
+                |ek_tpm| {
+                    self.update_tpmdata("ek_tpm", &base64::encode(&ek_tpm))
+                },
+            )?;
+            self.update_tpmdata("ek_pw", &ek_pw)?;
+            self.write_tpmdata()
+        })?;
+        Ok(())
+    }
+
+    fn get_pub_ek(&mut self) -> Result<(), KeylimeTpmError> {
+        let tf = NamedTempFile::new()?;
+        let tf_path = self.temp_file_get_path(&tf)?;
+
+        let cmd = match self.legacy_tools {
+            true => format!(
+                "tpm2_readpublic -H {} -o {} -f pem",
+                hex::encode(self.get_tpmdata("ek_handle")),
+                tf_path
+            ),
+            false => format!(
+                "tpm2-readpublic -c {} -o {} -f pem",
```

Comment:
needs an underscore

~~tpm2-readpublic~~

tpm2_readpublic

Reply:
I think I am going to design a test suite in an emulator environment.